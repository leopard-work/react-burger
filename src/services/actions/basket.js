export const ADD_TO_BASKET = 'ADD_TO_BASKET';
export const REMOVE_FROM_BASKET = 'REMOVE_FROM_BASKET';
export const SORT_BASKET = 'SORT_BASKET';
export const BASKET_CLEAR = 'BASKET_CLEAR';