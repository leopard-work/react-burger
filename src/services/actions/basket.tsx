import { ItemProps } from "../../utils/types";

export const ADD_TO_BASKET: "ADD_TO_BASKET" = "ADD_TO_BASKET";
export const REMOVE_FROM_BASKET: "REMOVE_FROM_BASKET" = "REMOVE_FROM_BASKET";
export const SORT_BASKET: "SORT_BASKET" = "SORT_BASKET";
export const BASKET_CLEAR: "BASKET_CLEAR" = "BASKET_CLEAR";

interface ADD_TO_BASKET_ACTION {
  readonly type: typeof ADD_TO_BASKET;
  item: ItemProps;
}
interface REMOVE_FROM_BASKET_ACTION {
  readonly type: typeof REMOVE_FROM_BASKET;
  item: ItemProps & { uuid: string };
}
interface SORT_BASKET_ACTION {
  readonly type: typeof SORT_BASKET;
  index: number;
}
export interface BASKET_CLEAR_ACTION {
  readonly type: typeof BASKET_CLEAR;
}

export type BASKET_ACTIONS =
  | ADD_TO_BASKET_ACTION
  | REMOVE_FROM_BASKET_ACTION
  | SORT_BASKET_ACTION
  | BASKET_CLEAR_ACTION;
