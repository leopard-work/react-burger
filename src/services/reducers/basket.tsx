import {
  ADD_TO_BASKET,
  BASKET_CLEAR,
  REMOVE_FROM_BASKET,
  SORT_BASKET,
} from "../actions/basket";

const initialState = {
  basket: [], // Корзина
};

export const basketReducer = (state = initialState, action: any) => {
  switch (action.type) {
    case ADD_TO_BASKET: {
      // Добавление товара в корзину
      if (action.item.type === "bun") {
        // Проверка булка или нет
        let itemFromBasket = state.basket.find((i: any) => i.type === "bun");
        if (itemFromBasket) {
          // Если булка есть
          const tempBasket = state.basket.filter(
            (i: any) => i.type !== action.item.type
          );
          return {
            ...state,
            basket: [...tempBasket, { ...action.item, count: 1 }],
          };
        } else {
          // Если булки нет
          return {
            ...state,
            basket: [...state.basket, { ...action.item, count: 1 }],
          };
        }
      } else {
        // Если не булка
        return {
          ...state,
          basket: [...state.basket, { ...action.item, count: 1 }],
        };
        // const itemFromBasket = state.basket.find(i => i._id === action.item._id);   // Проверка существования товара в корзине
        // if (itemFromBasket) {                                                       // Если товар есть в корзине
        //     state.basket.map((item) => {                                            // Увеличиваем количество
        //         if (item._id === action.item._id) {
        //             item.count++;
        //         }
        //         return state;
        //     })
        //     return {
        //         ...state
        //     }
        // }
        // else {                                                                      // Если товара нет, добавляем в корзину c увеличением счетчика
        //     return {
        //         ...state,
        //         basket: [...state.basket, {...action.item, count: 1}]
        //     }
        // }
      }
    }
    case REMOVE_FROM_BASKET: {
      // Удаление товара из корзины
      return {
        ...state,
        basket: state.basket.filter(
          (item: any) => item.uuid !== action.item.uuid
        ),
      };
      // if (action.item.count > 1) {                                                    // Если количество >1 уменьшаем количество
      //     state.basket.map((item) => {
      //         if (item._id === action.item._id) {
      //             item.count--;
      //         }
      //         return state;
      //     });
      //     return {
      //         ...state
      //     }
      // } else {                                                                        // Удаляем товар
      //     return {
      //         ...state,
      //         basket: state.basket.filter(item => item._id !== action.item._id)
      //     }
      // }
    }
    case SORT_BASKET: {
      // Сортировка корзины в зависимости от перемещенных элементов
      let tempBasket = [...state.basket];
      tempBasket[action.index] = tempBasket.splice(
        // @ts-ignore
        state.basket.indexOf(action.item),
        1,
        tempBasket[action.index]
      )[0];
      return {
        ...state,
        basket: tempBasket,
      };
    }
    case BASKET_CLEAR: {
      return {
        ...state,
        basket: [],
      };
    }
    default: {
      return state;
    }
  }
};
