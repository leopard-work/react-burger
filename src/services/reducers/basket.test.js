import {
  ADD_TO_BASKET,
  BASKET_ACTIONS,
  BASKET_CLEAR,
  REMOVE_FROM_BASKET,
  SORT_BASKET,
} from "../actions/basket";

import { initialState, basketReducer } from "./basket";
import renderer from "react-test-renderer";

import React from "react";

const testItem = {
  _id: "",
  name: "testItem",
  type: "",
  proteins: 0,
  fat: 0,
  carbohydrates: 0,
  calories: 0,
  price: 0,
  image: "",
  image_mobile: "",
  image_large: "",
  __v: 0,
  count: 0,
};
const testItem2 = { ...testItem, name: "testItem 2" };

describe("Корзина", () => {
  it("- исходное состояние", () => {
    expect(basketReducer(undefined, {})).toEqual(initialState);
    const tree = renderer.create(<>Исходное состояние</>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("- обработка ADD_TO_BASKET", () => {
    expect(
      basketReducer(initialState, {
        type: ADD_TO_BASKET,
        item: testItem,
      })
    ).toEqual({
      ...initialState,
      basket: [...initialState.basket, { ...testItem, count: 1 }],
    });
    const tree = renderer.create(<>Добавление товара в корзину</>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("- обработка REMOVE_FROM_BASKET", () => {
    expect(
      basketReducer(
        { ...initialState, basket: [testItem] },
        {
          type: REMOVE_FROM_BASKET,
          item: testItem,
        }
      )
    ).toEqual({
      ...initialState,
      basket: [],
    });
    const tree = renderer.create(<>Удаление товара из корзины</>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("- обработка SORT_BASKET", () => {
    expect(
      basketReducer(
        { ...initialState, basket: [testItem, testItem2] },
        {
          type: SORT_BASKET,
          item: testItem,
          index: 1,
        }
      )
    ).toEqual({
      ...initialState,
      basket: [testItem2, testItem],
    });
    const tree = renderer
      .create(<>Сортировка корзины в зависимости от перемещенных элементов</>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("- обработка BASKET_CLEAR", () => {
    expect(
      basketReducer(
        { ...initialState, basket: [testItem] },
        {
          type: BASKET_CLEAR,
        }
      )
    ).toEqual({
      ...initialState,
      basket: [],
    });
    const tree = renderer.create(<>Чистка корзины</>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

// SORT_BASKET
