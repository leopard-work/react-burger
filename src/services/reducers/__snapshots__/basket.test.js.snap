// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Корзина - исходное состояние 1`] = `"Исходное состояние"`;

exports[`Корзина - обработка ADD_TO_BASKET 1`] = `"Добавление товара в корзину"`;

exports[`Корзина - обработка BASKET_CLEAR 1`] = `"Чистка корзины"`;

exports[`Корзина - обработка REMOVE_FROM_BASKET 1`] = `"Удаление товара из корзины"`;

exports[`Корзина - обработка SORT_BASKET 1`] = `"Сортировка корзины в зависимости от перемещенных элементов"`;
